exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/SqlPatternContainsMatcher.java

```diff
@@ -17,37 +17,166 @@
  */
 package org.apache.drill.exec.expr.fn.impl;
 
-public class SqlPatternContainsMatcher implements SqlPatternMatcher {
+public final class SqlPatternContainsMatcher implements SqlPatternMatcher {
   final String patternString;
   CharSequence charSequenceWrapper;
   final int patternLength;
+  final MatcherFcn matcherFcn;
 
   public SqlPatternContainsMatcher(String patternString, CharSequence charSequenceWrapper) {
-    this.patternString = patternString;
+    this.patternString       = patternString;
     this.charSequenceWrapper = charSequenceWrapper;
-    patternLength = patternString.length();
+    patternLength            = patternString.length();
+
+    // The idea is to write loops with simple condition checks to allow the Java Hotspot achieve
+    // better optimizations (especially vectorization)
+    if (patternLength == 1) {
+      matcherFcn = new Matcher1();
```

Comment:
Padma, I have two reasons to follow the added complexity
1) The new code is encapsulated within the Contains matching logic; doesn't increase code complexity
2) 
o I created a test with the original match logic, pattern and input were Strings though passed as CharSequence
o Ran the test with the new and old method (1 billion iterations) on MacOS
o pattern length 
o The old match method performed in 43sec where as the new one performed in 15sec
o The reason for the speedup is the custom matcher functions have less instructions (load and comparison)

Reply:
how does  matcherN perform compared to matcher1, matcher2, matcher3 for pattern lengths 1, 2 and 3 ? If matcherN performs well for patternLengths 1, 2 and 3, we can just have one matcher instead of multiple for different pattern lengths.

Reply:
I ran two types of tests to evaluate the generic vs custom method:
Test1 - A match does exist
o Custom method is 3x faster because the code will go to the "else" part
o A nested loop is always slower than unrolled code (the loop is also correlated with the outer one); please refer to this article (https://en.wikipedia.org/wiki/Loop_unrolling) on the benefits of loop unrolling
o Older match function performed in 59sec

Test2- A match doesn't exist
o Custom method and generic one perform in 15sec; this is because both perform a comparison and proceed to the next iteration
o Older match function performed in 45sec